pkg/rules/rules_test.go

```diff
@@ -1390,3 +1382,48 @@ func TestFilterRules(t *testing.T) {
 		})
 	}
 }
+func generateTestCases() []*rulespb.RuleGroup {
+	testCases := make([]*rulespb.RuleGroup, 0)
+
+	for i := 0; i < 100; i++ {
+		group := &rulespb.RuleGroup{
+			Name: "a",
+			File: "bar.yaml",
+			Rules: []*rulespb.Rule{
+				rulespb.NewRecordingRule(&rulespb.RecordingRule{Name: "a1"}),
+				rulespb.NewRecordingRule(&rulespb.RecordingRule{Name: "a2"}),
+			},
+		}
+
+		if i%2 == 0 {
+			group.Name = "b"
+			group.Rules = []*rulespb.Rule{
+				rulespb.NewRecordingRule(&rulespb.RecordingRule{Name: "b1"}),
+				rulespb.NewRecordingRule(&rulespb.RecordingRule{Name: "b2"}),
+			}
+		} else if i%3 == 0 {
+			group.Name = "c"
+			group.Rules = nil
+		}
+
+		testCases = append(testCases, group)
+	}
+
+	return testCases
+}
+func BenchmarkDedupGroups(b *testing.B) {
+
+	testCases := generateTestCases()
+	// Prepare some sample input data for the benchmark
+	groups := make([]*rulespb.RuleGroup, len(testCases))
+	copy(groups, testCases)
+	// Run the benchmark
+	b.ResetTimer()
+	for i := 0; i < 6; i++ {
```

Comment:
This is probably the reason that the benchmark results aren't being accurate. This loop should run till `b.N`. You can read more about this here https://dave.cheney.net/2013/06/30/how-to-write-benchmarks-in-go
```suggestion
	for i := 0; i < b.N; i++ {
```

Reply:
I tried with b.N but the original dedupGroups function that uses the sort but doesnt run to completion, maybe I was missing something. I used a small number instead to see if it ever runs

Reply:
Hmm, this works for me on the old dedupGroups method,
```go
func BenchmarkDedupGroups(b *testing.B) {
	testCases := generateTestCases()
	// Run the benchmark
	b.ResetTimer()
	b.ReportAllocs()
	for i := 0; i < b.N; i++ {
		b.StopTimer()
		groups := make([]*rulespb.RuleGroup, len(testCases))
		copy(groups, testCases)
		b.StartTimer()
		// Call the function under test
		_ = dedupGroups(groups)

	}
}
```

Outputs this,
```bash
➜ go test -v -benchmem -run=^$ -count=5 -bench ^BenchmarkDedupGroups$ github.com/thanos-io/thanos/pkg/rules           
goos: linux
goarch: amd64
pkg: github.com/thanos-io/thanos/pkg/rules
cpu: Intel(R) Core(TM) i7-10610U CPU @ 1.80GHz
BenchmarkDedupGroups
BenchmarkDedupGroups-8   	  56402	    21560 ns/op	   7365 B/op	      2 allocs/op
BenchmarkDedupGroups-8   	  57015	    21474 ns/op	   7286 B/op	      2 allocs/op
BenchmarkDedupGroups-8   	  54897	    21931 ns/op	   7565 B/op	      2 allocs/op
BenchmarkDedupGroups-8   	  49572	    21816 ns/op	   6703 B/op	      2 allocs/op
BenchmarkDedupGroups-8   	  51048	    21851 ns/op	   7143 B/op	      2 allocs/op
PASS
ok  	github.com/thanos-io/thanos/pkg/rules	11.263s
```